docs/analyzers/FixAllProvider.md

```diff
@@ -0,0 +1,109 @@
+ - Ensure that all the code actions registered by your code fixer have a non-null equivalence key. See the following section to determine how to select an equivalence key.
+
+Selecting an Equivalence key for code actions
+=============================================
+
+Each unique equivalence key for a code fixer defines a unique equivalence class of code actions. Equivalence key of the trigger code action is part of the `FixAllContext` and is used to determine the FixAll occurrences code fix.
+Normally, you can use the **'title'** of the code action as the equivalence key. However, there are cases where you may desire to have different values. Let us take an example to get a better understanding.
+
+Let us consider the [C# SimplifyTypeNamesCodeFixProvider](http://source.roslyn.io/#q=Microsoft.CodeAnalysis.CSharp.CodeFixes.SimplifyTypeNames.SimplifyTypeNamesCodeFixProvider) that registers multiple code actions and also has FixAll support. This code fixer offers fixes to simplify the following expressions:
+ - This expressions of the form 'this.x' to 'x'.
+ - Qualified type names of the form 'A.B' to 'B'.
+ - Member access expressions of the form 'A.M' to 'M'.
+
+This fixer needs the following semantics for the corresponding FixAll occurrences code fixes:
+ - This expression simplification: Fix all should simplify all this expressions, regardless of the member being accessed (this.x, this.y, this.z, etc.).
+ - Qualified type name simplification: Fix all should simplify all qualified type names 'A.B' to 'B'. However, we don't want to simplify **all** qualified type names, such as 'C.D', 'E.F', etc. as that would be too generic a fix, which is not likely intended by the user.
+ - Member access expressions: Fix all should simplify all member access expressions 'A.M' to 'M'.
+
+It uses the below equivalence keys for it's registered code actions to get the desired FixAll behavior:
+ - This expression simplification: Generic resource string "Simplify this expression", which explicitly excludes the contents of the node being simplified.
+ - Qualified type name simplification: Formatted resource string "Simplify type name A.B", which explicitly includes the contents of the node being simplified.
+ - Member access expressions: Formatted resource string "Simplify type name A.M", which explicitly includes the contents of the node being simplified.
+
+Note that 'this expression simplification' fix requires a different kind of an equivalence class, then the other two simplifications. See method [GetCodeActionId](http://source.roslyn.io/Microsoft.CodeAnalysis.CSharp.Features/R/917a728e9783562f.html) for the actual implementation.
+
+To summarize, use the equivalence key that best suits the category of fixes to be applied as part of a FixAll operation.
+
+Built-in FixAllProvider
+=======================
```

Comment:
You might want to talk about the spectrum of fixallproviders. Something like:

When multiple fixes need to be applied to documents, there are various way to do it:
- **Sequential approach**: One way to do it is to compute diagnostics, pick one, ask a fixer to produce a codeaction to fix that, apply it. Now for the resulting new compilation, recomputed diagnostics, pick the next one and repeat the process. This approach would be very slow but would lead to correct results (unless it doesn't converge where one fix introduces a diagnostic that was just fixed by a previous fix). We chose to not implement this approach.
- **Batch fix approach** - Another way to do this to compute all the diagnostics, pick each diagnostic and give it to a fixer to and ask it apply it to produce a new solution. If there were 'n' diagnostics, there would be 'n' new solutions. Now just merge them all together in one go. This may produce incorrect results (when different fixes change the same region of code in different ways) but it is very fast. We have one implementation of this approach in `WellKnownFixAllProviders.BatchFixer`
- **Custom approach** - Depending on the fix, there may be a custom solution to fix multiple issues. For example, consider an analyzer that simply needs to generate one file as the fix for any instance of the issue. Instead of generating the same file over and over using the previous two approaches, one could write a custom `FixAllProvider` that simply generates the file once if there were any diagnostics at all.

Since there are various ways of fixing all issues, we've implemented a framework and provided the one general implementation that we think is useful in many cases.